### ISSUE TITLE
Docker Compose Startup fails with plethora of errors

### ISSUE THREAD

**Issue by invisible-walrus** [TandoorRecipes/recipes#1032]
[_recipes_web_recipes_1_logs.txt](https://github.com/TandoorRecipes/recipes/files/7489019/_recipes_web_recipes_1_logs.txt)
[_recipes_nginx_recipes_1_logs.txt](https://github.com/TandoorRecipes/recipes/files/7489020/_recipes_nginx_recipes_1_logs.txt)
[_recipes_db_recipes_1_logs.txt](https://github.com/TandoorRecipes/recipes/files/7489021/_recipes_db_recipes_1_logs.txt)

When I start up a fresh install on my docker compose instance, it returns all kinds of errors across all three containers. I only change the secret and the postgres password, as well as changing the ngnix port to 2080, to avoid conflicts, but the webserver returns a 502, the db announces that "the server must be started by the user that owns the data directory" (tried as sudo, root user and my default user) and the web container announces "worker failed to boot" as well as "could not translate host name "db_recipes" to address: Try again"

**Comment by invisible-walrus** [comment]
I have tried: removing all folders created by the compose file, two different systems (ubuntu 20.04, dietpi [debian]), waiting a while, reinstalling everything many times, default everything. no change to errors.

**Comment by smilerz** [comment]
`FATAL:  data directory "/var/lib/postgresql/data" has wrong ownership`

the folder you have that mounted to must be owned by the same user that is running docker.  (probably root?)

**Comment by invisible-walrus** [comment]
Where is that folder located? there are no staticfiles generated in any of the volumes that the compose creates, and no folder matching that name in the host /var/lib/

**Comment by smilerz** [comment]
Can you post your `docker-compose.yml` with secrets/passwords removed?

**Comment by invisible-walrus** [comment]
version: "3"
services:
  db_recipes:
    restart: always
    image: postgres:11-alpine
    volumes:
      - ./postgresql:/var/lib/postgresql/data
    env_file:
      - ./.env

  web_recipes:
    image: vabene1111/recipes
    restart: always
    env_file:
      - ./.env
    volumes:
      - staticfiles:/opt/recipes/staticfiles
      - nginx_config:/opt/recipes/nginx/conf.d
      - ./mediafiles:/opt/recipes/mediafiles
    depends_on:
      - db_recipes

  nginx_recipes:
    image: nginx:mainline-alpine
    restart: always
    ports:
      - 2080:80
    env_file:
      - ./.env
    depends_on:
      - web_recipes
    volumes:
      - nginx_config:/etc/nginx/conf.d:ro
      - staticfiles:/static
      - ./mediafiles:/media

volumes:
  nginx_config:
  staticfiles:

**Comment by invisible-walrus** [comment]
and my .env file:

# only set this to true when testing/debugging
# when unset: 1 (true) - dont unset this, just for development
DEBUG=0
SQL_DEBUG=0

# hosts the application can run under e.g. recipes.mydomain.com,cooking.mydomain.com,...
ALLOWED_HOSTS=*

# random secret key, use for example `base64 /dev/urandom | head -c50` to generate one
SECRET_KEY=[REDACTED]

# your default timezone See https://timezonedb.com/time-zones for a list of timezones
TIMEZONE=Europe/Berlin

# add only a database password if you want to run with the default postgres, otherwise change settings accordingly
DB_ENGINE=django.db.backends.postgresql
# DB_OPTIONS= {} # e.g. {"sslmode":"require"} to enable ssl
POSTGRES_HOST=db_recipes
POSTGRES_PORT=5432
POSTGRES_USER=djangouser
POSTGRES_PASSWORD=password
POSTGRES_DB=djangodb

# database connection string, when used overrides other database settings.
# format might vary depending on backend
# DATABASE_URL = engine://username:password@host:port/dbname

# the default value for the user preference 'fractions' (enable/disable fraction support)
# default: disabled=0
FRACTION_PREF_DEFAULT=0

# the default value for the user preference 'comments' (enable/disable commenting system)
# default comments enabled=1
COMMENT_PREF_DEFAULT=1

# Users can set a amount of time after which the shopping list is refreshed when they are in viewing mode
# This is the minimum interval users can set. Setting this to low will allow users to refresh very frequently which
# might cause high load on the server. (Technically they can obviously refresh as often as they want with their own scripts)
SHOPPING_MIN_AUTOSYNC_INTERVAL=5

# Default for user setting sticky navbar
# STICKY_NAV_PREF_DEFAULT=1

# If base URL is something other than just /  (you are serving a subfolder in your proxy for instance http://recipe_app/recipes/)
# SCRIPT_NAME=/recipes

# If staticfiles are stored at a different location uncomment and change accordingly, MUST END IN /
# this is not required if you are just using a subfolder
# This can either be a relative path from the applications base path or the url of an external host
# STATIC_URL=/static/

# If mediafiles are stored at a different location uncomment and change accordingly, MUST END IN /
# this is not required if you are just using a subfolder
# This can either be a relative path from the applications base path or the url of an external host
# MEDIA_URL=/media/

# Serve mediafiles directly using gunicorn. Basically everyone recommends not doing this. Please use any of the examples
# provided that include an additional nxginx container to handle media file serving.
# If you know what you are doing turn this back on (1) to serve media files using djangos serve() method.
# when unset: 1 (true) - this is temporary until an appropriate amount of time has passed for everyone to migrate
GUNICORN_MEDIA=0

# S3 Media settings: store mediafiles in s3 or any compatible storage backend (e.g. minio)
# as long as S3_ACCESS_KEY is not set S3 features are disabled
# S3_ACCESS_KEY=
# S3_SECRET_ACCESS_KEY=
# S3_BUCKET_NAME=
# S3_REGION_NAME= # default none, set your region might be required
# S3_QUERYSTRING_AUTH=1 # default true, set to 0 to serve media from a public bucket without signed urls
# S3_QUERYSTRING_EXPIRE=3600 # number of seconds querystring are valid for
# S3_ENDPOINT_URL= # when using a custom endpoint like minio

# Email Settings, see https://docs.djangoproject.com/en/3.2/ref/settings/#email-host
# Required for email confirmation and password reset (automatically activates if host is set)
# EMAIL_HOST=
# EMAIL_PORT=
# EMAIL_HOST_USER=
# EMAIL_HOST_PASSWORD=
# EMAIL_USE_TLS=0
# EMAIL_USE_SSL=0
# DEFAULT_FROM_EMAIL= # email sender address (default 'webmaster@localhost')
# ACCOUNT_EMAIL_SUBJECT_PREFIX= # prefix used for account related emails (default "[Tandoor Recipes] ")

# allow authentication via reverse proxy (e.g. authelia), leave off if you dont know what you are doing
# see docs for more information https://vabene1111.github.io/recipes/features/authentication/
# when unset: 0 (false)
REVERSE_PROXY_AUTH=0

# Default settings for spaces, apply per space and can be changed in the admin view
# SPACE_DEFAULT_MAX_RECIPES=0 # 0=unlimited recipes
# SPACE_DEFAULT_MAX_USERS=0 # 0=unlimited users per space
# SPACE_DEFAULT_MAX_FILES=0 # Maximum file storage for space in MB. 0 for unlimited, -1 to disable file upload.
# SPACE_DEFAULT_ALLOW_SHARING=1 # Allow users to share recipes with public links

# allow people to create accounts on your application instance (without an invite link)
# when unset: 0 (false)
# ENABLE_SIGNUP=0

# If signup is enabled you might want to add a captcha to it to prevent spam
# HCAPTCHA_SITEKEY=
# HCAPTCHA_SECRET=

# if signup is enabled you might want to provide urls to data protection policies or terms and conditions
# TERMS_URL=
# PRIVACY_URL=
# IMPRINT_URL=

# enable serving of prometheus metrics under the /metrics path
# ATTENTION: view is not secured (as per the prometheus default way) so make sure to secure it
# trough your web server (or leave it open of you dont care if the stats are exposed)
# ENABLE_METRICS=0

# allows you to setup OAuth providers
# see docs for more information https://vabene1111.github.io/recipes/features/authentication/
# SOCIAL_PROVIDERS = allauth.socialaccount.providers.github, allauth.socialaccount.providers.nextcloud,

# Should a newly created user from a social provider get assigned to the default space and given permission by default ?
# ATTENTION: This feature might be deprecated in favor of a space join and public viewing system in the future
# default 0 (false), when 1 (true) users will be assigned space and group
# SOCIAL_DEFAULT_ACCESS = 1

# if SOCIAL_DEFAULT_ACCESS is used, which group should be added
# SOCIAL_DEFAULT_GROUP=guest

# Django session cookie settings. Can be changed to allow a single django application to authenticate several applications
# when running under the same database
# SESSION_COOKIE_DOMAIN=.example.com
# SESSION_COOKIE_NAME=sessionid # use this only to not interfere with non unified django applications under the same top level domain

# by default SORT_TREE_BY_NAME is disabled this will store all Keywords and Food in the order they are created
# enabling this setting makes saving new keywords and foods very slow, which doesn't matter in most usecases.
# however, when doing large imports of recipes that will create new objects, can increase total run time by 10-15x
# Keywords and Food can be manually sorted by name in Admin
# This value can also be temporarily changed in Admin, it will revert the next time the application is started
# This will be fixed/changed in the future by changing the implementation or finding a better workaround for sorting
# SORT_TREE_BY_NAME=0
# LDAP authentication
# default 0 (false), when 1 (true) list of allowed users will be fetched from LDAP server
#LDAP_AUTH=
#AUTH_LDAP_SERVER_URI=
#AUTH_LDAP_BIND_DN=
#AUTH_LDAP_BIND_PASSWORD=
#AUTH_LDAP_USER_SEARCH_BASE_DN=

**Comment by invisible-walrus** [comment]
[env.txt](https://github.com/TandoorRecipes/recipes/files/7489117/env.txt)

and my .env

**Comment by smilerz** [comment]
the postgres folder is `./postgresql` which I think is in your home folder(?)

**Comment by invisible-walrus** [comment]
there is a postgres folder there, but nothing inside it. i run the docker-compose up -d as root, then when that didnt work i chown'd the folder as root too, no change.

**Comment by invisible-walrus** [comment]
i have now tried removing the ./postgresql & ./mediafiles folders and recreated them using sudo mkdir. no change. added a ./postgresql/data directory with sudo mkdir for s&g, no change.

**Comment by smilerz** [comment]
can you confirm that root has rw permission as well?

**Comment by invisible-walrus** [comment]
results of a `namei -l
`
drwxr-xr-x root root /
drwxrwxrwx root root Honor (drive name)
drwxrwxrwx root root Recipes
drwxrwxrwx root root postgresql

**Comment by smilerz** [comment]
I'm honestly not 100% where `.` is in terms of docker.  I created folders in `/mnt` and modified the yml to point there instead.

**Comment by invisible-walrus** [comment]
its just the current working directory. /Honor/Recipes/ is where i keep the .env and docker-compose.yml. the only other things in that folder are the /mediafiles & /postgresql directories

**Comment by smilerz** [comment]
I know its current working directory - I just don't know what the docker process thinks is the current working directory.  The postgres container thinks it doesn't have ownership - which suggests maybe it isn't where I thought it was.

**Comment by invisible-walrus** [comment]
ah gotcha. i tried updating the docker-compose with the full path, no change. there are no other folders named postgresql on my system either.

**Comment by smilerz** [comment]
I thought I was going to be able to provide a quick fix, but I don't know docker well enough to go any deeper than that.  It's a standard postgres container, you might be able to find a solution in the docker forums or stackoverflow.

**Comment by invisible-walrus** [comment]
thanks for your help anyways!

**Comment by steffen-lorenz** [comment]
I have the same issue when updating to 1.0.0
`
web_recipes_1    | django.db.utils.OperationalError: could not translate host name "db_recipes" to address: Try again
web_recipes_1    |
`

**Comment by smilerz** [comment]
Can you post the logs of the DB container?

**Comment by vabene1111** [comment]
that is very weird, we never had this issue before but now multiple people appear to be having it. I 100% agree with the debugging steps @smilerz took but the results (folder owned by root with 777 permission) definitely show that the container should be able to access the files. 

just for clarification the `./postgresdata:/folder/in/pg/container` is just relative to the docker compose file (this is a bind mount) while all volumes (the ones listed at the volumes section of the compose file and written as `volumename:/folder/in/container` are by default created in a docker controlled directory. 

Whats also interesting is that the docker folder permission error happens at the same time as the issue @steffen-lorenz mentioned which is from the topic something completly differnt (docker not being able to resolve a host name). 

One thing that i am not sure if even possible but asking for completeness sake: does your docker run as root ?

It also appears that the data directoy issue is not specific to tandoor but postgres and seems to happen to others as well, in this thread a workaround is proposed FATAL: data directory "/var/lib/postgresql/data" has wrong ownership

One thing i could think of you could try is store your compose file in an entirely different directory, maybe something higher up the tree causes issues, i usually use `/srv/recipes` for the installation, maybe that helps.

**Comment by steffen-lorenz** [comment]
Seems that the issue fixed itself over-night ... Now it runs again as expected.
When checking `docker-compose logs` I can see that docker(?) tried to start the worker multiple times (~190 times) and suddenly it starts working again.
And yes as far as I see Docker runs as root.
Anything I can provide to further analyse the issue?

**Comment by vabene1111** [comment]
Very strange. I guess since we never had this issue before we can ignore it for now, if it ever reappears we will need to look into why it happened and add something to the FAQ.

**Comment by invisible-walrus** [comment]
My issues never cleared up, it's still not working. Not really sure what to
do next

**Comment by Ruediger34** [comment]
I agree with @invisible-walrus 
Have tried to setup tandoor on a new install of ubuntu 20.04 LTS, ubuntu 21.10 and on a debian 11 system with docker-compose. 
On all systems it fails with 502 bad gateway.

Have tried to install it in /srv/recipes and in /opt/recipes all with root for testing but no success.

.env and docker-compose.yml with default settings from the repo.

docker-compose log shows the same error:
> web_recipes_1 | django.db.utils.OperationalError: could not translate host name "db_recipes" to address: Try again web_recipes_1

**Comment by vabene1111** [comment]
Hmm this is a problem. I am not sure, i remember having had the exact same issue before but for some reason it just dissappeard at some point. What the issue says is that the application is trying to connect to a host called `db_recipes` as specified in the `.env` file under `POSTGRES_HOST`. 
This should normally be possible as docker compose (or docker, not sure) makes the names you specify in the compose file available as host names for other containers (same goes for the nginx which connects to `web_recipes`).

I dont know why this host name is not working in your setup, previously this was related to people renaming the containers (which you did not do) or splitting the compose file into multiple files (which you also did not do). 

I could think of a few more things: is the DB container running, are you using the lastest docker and compose versions, do you have any special docker network settings set ? but other than that i am not really sure why this happens. 

What you can do is assign a fixed IP to the docker container (i think at least, have never done so) and then use that as the database host.

**Comment by smilerz** [comment]
I had this issue as well when I tried to get it working on synology to look at a separate issue, but haven’t had time to figure it out.  It’s almost certainly a docker issue.  Stackoverflow might be of assistance.

**Comment by neptuniux** [comment]
I've tried to setup the project and I'm having the same problems, with the last version of docker and docker-compose. I've ran everything in sudo on a clean installed ubnutu and Raspbian.

**Comment by vabene1111** [comment]
do you mean by the same problem the exact same error message `web_recipes_1 | django.db.utils.OperationalError: could not translate host name "db_recipes" to address: Try again`

**Comment by reprz** [comment]
I have also difficulties setting up this container: 

django.db.utils.operationalerror: could not translate host name "db_recipes" to address: name does not resolve

This error happens all the time no matter how I try to setup container. I tried with docker CLI and Docker compose. No luck.

**Comment by invisible-walrus** [comment]
could you post the full compose? im sure people in the future would appreciate having this

**Comment by vabene1111** [comment]
I strongly discourage anyone from using this compose !! This does not include the nginx container and therefor is not one of the recommended setups. There is a reason this container exists and I will at some point stop giving support to setups like this.

**Comment by reprz** [comment]
Well I just tried to say that I cannot get this container working in any other way and it looks like there is no resolution to this issue either. Closing issue is not a way to resolve this though.

**Comment by vabene1111** [comment]
i am sorry reprz, we had hundreds of people installing tandoor over the last few days, most of them without any problems. I have personally installed it on many servers and it works fine if following the recommended setup guides. Even tough i do not have the time to i try to help everyone but if 5 people write in one issue, each providing only parts of the information that i explicitly list as required for me to help you then i simply loose track of what needs to be done. 

Can you please open your own issue, provide all the information the template asks you for and i will see if i can help you. Other than that please remember that I and all other contributers are doing this on their free time, the setup is proven to work and if anyone helps you at all getting this to work for you thanking them for their effort would probably be the least to do.

**Comment by reprz** [comment]
I am fairly new to using containers so there's a big possibility that I (and others) have failed configuring the container. I was just wondering what could be wrong as I followed installation guide and cannot get it working with nginx but it works without it. But thanks for answering.

**Comment by vabene1111** [comment]
It is really hard for me to understand what is wrong for you from my standpoint. I need certain information, mainly the compose, logs, environment and some Information about the OS. If you post that in a single issue i can look at whats wrong.
